### PR TITLE
Feature/sample go

### PR DESCRIPTION
**Pull request by Shylock-Hg**
Sub job of #2602

### CODE REVIEW

**Comment by nevermore3** [comment]
i think you should write a design document first，
And  this design does not sample every step under many conditions

**Comment by Shylock-Hg** [comment]
> i think you should write a design document first，
> And this design does not sample every step under many conditions

Multiple steps need change some plan nodes, I plan to do it later. Now just one step.

**Comment by nevermore3** [comment]
> > i think you should write a design document first，
> > And this design does not sample every step under many conditions
> 
> Multiple steps need change some plan nodes, I plan to do it later. Now just one step.

it only  work for  go from vid over * yield $^.tag.name, edge.name sample[1], 

It won't work with a little change, for example:  go  from vid over * yield $$.tag.name,  go from vid over * yield edge._dst as id | go from $-.id over * yield edge.name sample[1], ....

**Comment by Shylock-Hg** [comment]
> > > i think you should write a design document first，
> > > And this design does not sample every step under many conditions
> > 
> > 
> > Multiple steps need change some plan nodes, I plan to do it later. Now just one step.
> 
> it only work for go from vid over * yield $^.tag.name, edge.name sample[1],
> 
> It won't work with a little change, for example: go from vid over * yield $$.tag.name, go from vid over * yield edge._dst as id | go from $-.id over * yield edge.name sample[1], ....

With input, now sample the result of join. Do you think it's problem?

File: tests/tck/features/optimizer/PushSampleDownRule.feature [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      GO 1 STEPS FROM "James Harden" OVER like REVERSELY LIMIT [2]
+      """
+    Then the result should be, in any order:
+      | like._dst         |
+      | "Dejounte Murray" |
+      | "Luka Doncic"     |
+    And the execution plan should be:
+      | id | name         | dependencies | operator info  |
+      | 5  | Project      | 6            |                |
+      | 6  | Limit        | 7            |                |
```
Comment (czpmango) [inline tests/tck/features/optimizer/PushSampleDownRule.feature:22]: Can the limit node be deleted(bcz it's being pushed down)？

Comment (Shylock-Hg) [reply tests/tck/features/optimizer/PushSampleDownRule.feature:22]: Push down only perform in single storage node, when graph collect the many results from different storage we need limit again.

Comment (czpmango) [reply tests/tck/features/optimizer/PushSampleDownRule.feature:22]: ACK.

File: tests/tck/features/go/SampleLimit.intVid.feature [hunk 2]
```diff
@@ -0,0 +1,65 @@
+# Copyright (c) 2021 vesoft inc. All rights reserved.
+#
+# This source code is licensed under Apache 2.0 License,
+# attached with Common Clause Condition 1.0, found in the LICENSES directory.
+Feature: Sample and limit
```
Comment (czpmango) [inline tests/tck/features/go/SampleLimit.intVid.feature:5]: How does this work in match statement？Add
 some cases?

File: tests/tck/features/optimizer/PushSampleDownRule.feature [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      | 7  | GetNeighbors | 0            | {"limit": "2"} |
+      | 0  | Start        |              |                |
+    When profiling query:
+      """
+      GO 2 STEPS FROM "James Harden" OVER like REVERSELY LIMIT [2, 2]
```
Comment (czpmango) [inline tests/tck/features/optimizer/PushSampleDownRule.feature:27]: Is this statement very similar to `GO 2 STEPS FROM "James Harden" OVER like REVERSELY SAMPLE [2, 2] | LIMIT 2`? The only difference is the sampling algorithm, right?

File: tests/tck/features/optimizer/PushSampleDownRule.feature [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    And the execution plan should be:
+      | id | name         | dependencies | operator info                    |
+      | 1  | Project      | 2            |                                  |
+      | 2  | Sample       | 3            |                                  |
+      | 3  | GetNeighbors | 4            | {"limit": "2", "random": "true"} |
+      | 4  | Start        |              |                                  |
+    When profiling query:
+      """
+      GO 2 STEPS FROM "James Harden" OVER like REVERSELY SAMPLE [2, 2]
+      """
```
Comment (czpmango) [inline tests/tck/features/optimizer/PushSampleDownRule.feature:63]: Add this case:
```
profile GO 2 STEPS FROM "James Harden" OVER like REVERSELY SAMPLE [2, 3]
```

File: tests/tck/features/optimizer/PushSampleDownRule.feature [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      | 9  | Project      | 12           |                                                       |
+      | 12 | Sample       | 13           |                                                       |
+      | 13 | GetNeighbors | 6            | {"limit": "2", "random": "true"}                      |
+      | 6  | Loop         | 0            | {"loopBody": "5"}                                     |
+      | 5  | Dedup        | 4            |                                                       |
+      | 4  | Project      | 20           |                                                       |
+      | 20 | Sample       | 21           |                                                       |
+      | 21 | GetNeighbors | 1            | {"limit": "$__VAR_2[($__VAR_1-1)]", "random": "true"} |
+      | 1  | Start        |              |                                                       |
+      | 0  | Start        |              |                                                       |
```
Comment (czpmango) [inline tests/tck/features/optimizer/PushSampleDownRule.feature:79]: You could add some operator info to illustrate `"limit": "$__VAR_2[($__VAR_1-1)]"`, like where does the `$__VAR_2` and `($__VAR_1-1)` come from.

File: tests/tck/features/go/SampleLimit.intVid.feature (hunk 2 above)
Comment (Shylock-Hg) [reply tests/tck/features/go/SampleLimit.intVid.feature:5]: Don't support.

File: tests/tck/features/optimizer/PushSampleDownRule.feature [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+      | 20 | Limit        | 21           |                                                        |
+      | 21 | GetNeighbors | 1            | {"limit": "$__VAR_2[($__VAR_1-1)]", "random": "false"} |
+      | 1  | Start        |              |                                                        |
+      | 0  | Start        |              |                                                        |
+
+  Scenario: push sample down to GetNeighbors
+    When profiling query:
+      """
+      GO 1 STEPS FROM "James Harden" OVER like REVERSELY SAMPLE [2]
+      """
```
Comment (czpmango) [inline tests/tck/features/optimizer/PushSampleDownRule.feature:49]: How to handle property sampling?
```
GO 1 STEPS FROM "James Harden" OVER like REVERSELY YIELD $$.player.age AS age SAMPLE [2]
```

File: tests/tck/features/optimizer/PushSampleDownRule.feature (hunk 3 above)
Comment (Shylock-Hg) [reply tests/tck/features/optimizer/PushSampleDownRule.feature:27]: Not. Limit clause just perform the final result.

File: tests/tck/features/go/SampleLimit.intVid.feature (hunk 2 above)
Comment (czpmango) [reply tests/tck/features/go/SampleLimit.intVid.feature:5]: Only supported in go statement.

Comment (Shylock-Hg) [reply tests/tck/features/go/SampleLimit.intVid.feature:5]: Yes

File: tests/tck/features/go/SampleLimit.intVid.feature [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+  Background: Prepare space
```
Comment (czpmango) [inline tests/tck/features/go/SampleLimit.intVid.feature:7]: Add case:
```
$var=go from "Tim Duncan" over like yield like._dst as dst; go 2 steps from $var.dst over like yield $var.dst as dst1,like._dst as dst2 LIMIT [2,3]
```

File: tests/tck/features/optimizer/PushSampleDownRule.feature (hunk 3 above)
Comment (Shylock-Hg) [reply tests/tck/features/optimizer/PushSampleDownRule.feature:27]: And the GO step limit/sample perform on the result of GN

Comment (czpmango) [reply tests/tck/features/optimizer/PushSampleDownRule.feature:27]: I mean `sample[2,2]+limit 2` and `limit [2,2]`, Is there any specific difference in the execution plan between them?

File: tests/tck/features/optimizer/PushSampleDownRule.feature (hunk 4 above)
Comment (Shylock-Hg) [reply tests/tck/features/optimizer/PushSampleDownRule.feature:63]: The result of this query is not determined

File: tests/tck/features/optimizer/PushSampleDownRule.feature (hunk 5 above)
Comment (Shylock-Hg) [reply tests/tck/features/optimizer/PushSampleDownRule.feature:79]: It's just inner variable.

File: tests/tck/features/optimizer/PushSampleDownRule.feature (hunk 6 above)
Comment (Shylock-Hg) [reply tests/tck/features/optimizer/PushSampleDownRule.feature:72]: Add to test cases.

File: tests/tck/features/go/SampleLimit.intVid.feature (hunk 7 above)
Comment (Shylock-Hg) [reply tests/tck/features/go/SampleLimit.intVid.feature:7]: Add.

File: tests/tck/features/optimizer/PushSampleDownRule.feature (hunk 4 above)
Comment (czpmango) [reply tests/tck/features/optimizer/PushSampleDownRule.feature:122]: Just explain it.

Comment (Shylock-Hg) [reply tests/tck/features/optimizer/PushSampleDownRule.feature:122]: The explain is same with go 2 steps.

File: tests/tck/features/optimizer/PushSampleDownRule.feature (hunk 5 above)
Comment (czpmango) [reply tests/tck/features/optimizer/PushSampleDownRule.feature:138]: I get it.

File: src/graph/context/Iterator.cpp [hunk 8]
```diff
@@ -6,6 +6,9 @@
 
 #include "graph/context/Iterator.h"
 
+#include <cstdio>
```
Comment (czpmango) [inline src/graph/context/Iterator.cpp:9]: Is that necessary?

Comment (Shylock-Hg) [reply src/graph/context/Iterator.cpp:9]: Fixed.

**Review by nevermore3 — APPROVED** [review]

**Review by czpmango — DISMISSED** [review]

File: tests/tck/features/go/SampleLimit.feature [hunk 9]
```diff
@@ -0,0 +1,65 @@
+# Copyright (c) 2021 vesoft inc. All rights reserved.
+#
+# This source code is licensed under Apache 2.0 License,
+# attached with Common Clause Condition 1.0, found in the LICENSES directory.
+Feature: Sample and limit
+
+  Background: Prepare space
+    Given a graph with space named "nba"
+
+  Scenario: Sample Limit Go in One step
```
Comment (CPWstatic) [inline tests/tck/features/go/SampleLimit.feature:10]:
```
GO FROM 'Tim Duncan' OVER like LIMIT 1; 

GO FROM 'Tim Duncan' OVER like LIMIT ["1"];
```

Please add tests as above, also the sample clause.

File: src/graph/planner/ngql/GoPlanner.h [hunk 10]
```diff
@@ -64,13 +64,37 @@ class GoPlanner final : public Planner {
 
   PlanNode* buildOneStepJoinPlan(PlanNode* gn);
 
+  template <typename T>
+  PlanNode* buildSampleLimitImpl(PlanNode* input, T sampleLimit);
```
Comment (CPWstatic) [inline src/graph/planner/ngql/GoPlanner.h:68]: Why do we need generics here?

File: src/graph/planner/ngql/GoPlanner.cpp [hunk 11]
```diff
@@ -488,6 +540,10 @@ StatusOr<SubPlan> GoPlanner::transform(AstContext* astCtx) {
 
   auto& steps = goCtx_->steps;
   if (steps.isMToN()) {
+    // TODO(shylock) need read the limit number in runtime for the Loop
```
Comment (CPWstatic) [inline src/graph/planner/ngql/GoPlanner.cpp:543]: I thought you were done this?

File: src/graph/planner/ngql/GoPlanner.h (hunk 10 above)
Comment (Shylock-Hg) [reply src/graph/planner/ngql/GoPlanner.h:68]: Accept integer and expression.

File: src/graph/planner/ngql/GoPlanner.cpp (hunk 11 above)
Comment (Shylock-Hg) [reply src/graph/planner/ngql/GoPlanner.cpp]: Fixed.

File: tests/tck/features/go/SampleLimit.feature (hunk 9 above)
Comment (Shylock-Hg) [reply tests/tck/features/go/SampleLimit.feature:10]: Done

**Review by CPWstatic — APPROVED** [review]

**Review by nevermore3 — APPROVED** [review]

**Review by czpmango — APPROVED** [review]

**Pull request merged**